### DISCUSSION TITLE
PredictiveBackGesture slide animation not working trying to pop

### DISCUSSION THREAD

**Discussion by OsmanMusse** [arkivanov/Decompose#699] [Q&A] [answered]
Hello @arkivanov hope you're doing great. I've got an issue with the PredictiveBackGesture slide animation when I try to push a new screen it works, but when I try to slide it it won't allow me. It will only slide back when a button is clicked though.

Root Screen (App.kt)

    fun App(rootComponent: RootComponent) {
    val childStack = rootComponent.routerState

    Children(
        stack = childStack,
        modifier = Modifier.fillMaxSize(),
        content = { child ->
            println("Child Created == ")
           when(val childCreated = child.instance){
               is RootComponent.RootChild.LandingRoot -> Unit
               is RootComponent.RootChild.MainRoot -> MainScreen(childCreated.mainComponent)
           }
        }

    )}

Main Screen
 
     fun BottomNavigationView( component: HomeScreenComponent, scope: CoroutineScope,bottomSheetState:CupertinoSheetState) {

     NativeChildren(
         stack = component.fullScreenStack,
         modifier = Modifier.fillMaxSize(),
         animation = predictiveBackAnimation(
             backHandler = component.backHandler,
             onBack = {
                 println("DO SOMETHING")
             }
         ),
         backDispatcher = component.backDispatcher
     ) {
         when(it.instance) {
              is HomeScreenComponent.FullScreenChild.None -> {
                  CupertinoBottomSheetScaffold(
                      modifier = Modifier.fillMaxSize(),
                      scaffoldState = CupertinoBottomSheetScaffoldState(bottomSheetState),
                      bottomBar = {
                          val grayTitleColor = Color(158, 158, 158)
                          val grayIconColor = Color(190, 190, 190)
                          val primaryColor = Color(72, 134, 255)

                          val tabStack by component.tabStack.subscribeAsState()

                          CompositionLocalProvider(LocalRippleTheme provides NoRippleTheme) {
                              CupertinoNavigationBar(isTranslucent = false) {
                                  TabComponent.Tab.values().forEach { specificTab ->
                                      val (title, icon) = when (specificTab) {
                                          TabComponent.Tab.Home -> "Home" to MR.images.home_icon_lineal
                                          TabComponent.Tab.Account -> "Account" to MR.images.account_icon
                                          TabComponent.Tab.Post -> "Post" to MR.images.camera_icon_linear
                                          TabComponent.Tab.Saved -> "Saved" to MR.images.favourite_icon
                                          TabComponent.Tab.Settings -> "Settings" to MR.images.setting_icon_linear
                                      }

                                      val bottomTabsOffset = when (specificTab) {
                                          TabComponent.Tab.Account, TabComponent.Tab.Post -> 4.dp
                                          TabComponent.Tab.Saved, TabComponent.Tab.Settings -> 3.dp
                                          else -> 4.5.dp
                                      }

                                      NavigationBarItem(
                                          icon = {
                                              Icon(
                                                  painter = painterResource(icon),
                                                  contentDescription = null,
                                                  modifier = Modifier.fillMaxSize(if (specificTab == TabComponent.Tab.Saved) 0.45F else 0.50F)
                                                      .offset(y = bottomTabsOffset),
                                                  tint = Color.Unspecified,
                                              )
                                          },
                                          onClick = {
                                              component.onTabSelected(tab = specificTab)
                                          },
                                          selected = false,
                                          label = {
                                              Text(
                                                  text = "${title}",
                                                  fontSize = 12.5.sp,
                                                  color = if (tabStack.active.instance.tab == specificTab) primaryColor else grayTitleColor
                                              )
                                          }
                                      )

                                  }

                              }
                          }
                      },
                      sheetDragHandle = {},
                      sheetSwipeEnabled = false,
                      sheetContent = {
                          CupertinoBottomSheetContent {
                              CupertinoButton(
                                  modifier = Modifier.padding(top = 50.dp),
                                  onClick = {
                                      scope.launch {
                                          bottomSheetState.hide()
                                      }

                                  }
                              ) {
                                  Text(
                                      text = "Hello world",
                                  )
                              }
                          }

                      },
                  ) {
                      Children(
                          stack = component.tabStack,
                          modifier = Modifier,
                      ) { child ->
                          TabView(child.instance)
                      }
                  }
              }
              is HomeScreenComponent.FullScreenChild.PostScreen ->
                  CupertinoScaffold(modifier = Modifier.fillMaxSize().background(Color.Blue), hasNavigationTitle = true){
                      CupertinoButton(
                          modifier = Modifier.offset(y = 50.dp),
                          onClick = {
                              println("BACK BUTTON CLICKED ===")
                              component.onBack()
                          }
                      ){
                          Text("Go Back",color = Color.White)
                      }
                  }

              }
         }

     }

The video below demonstrates

https://github.com/arkivanov/Decompose/assets/38444091/db6572ac-93e3-45f2-8c64-1bfbd437bf0d

**Comment by arkivanov** [comment]
Might be a silly question, but what platform is it? Also it would be nice to see `HomeScreenComponent`.

**Reply by OsmanMusse** [reply]
compose Multiplatform, the screen recording is the app in IOS. using compose Cupertino for the predictiveBackAnimation function




HomeScreenComponent


    class HomeScreenComponent(
    private val tabFactory: TabComponent.Factory,
    private val componentContext: ComponentContext,
    private val onNavigateTo: (String) -> Unit): ComponentContext by componentContext  {

    init {
        println("HI THE H COMPONENT HAS BEEN CREATED")
    }


    private val _state = MutableValue(HomeScreenState())
    val state = _state


    private val mainNavigation = StackNavigation<MainConfig>()
    private val tabNavigation = StackNavigation<TabConfig>()

    val backDispatcher : BackDispatcher = (backHandler as? BackDispatcher) ?: BackDispatcher()


    private val _fullscreenStack =
        childStack(
            source = mainNavigation,
            initialConfiguration = MainConfig.None,
            serializer = MainConfig.serializer(),
            handleBackButton = true,
            childFactory = ::createFullScreenChild
        )
    private val _tabStack =
        childStack(
            source = tabNavigation,
            initialConfiguration = TabConfig(TabComponent.Tab.Home),
            key = "tabs",
            handleBackButton = true,
            childFactory = ::createTab,
        )


    val fullScreenStack: Value<ChildStack<*, FullScreenChild>> get() = _fullscreenStack
    val tabStack: Value<ChildStack<*, TabComponent>> get() = _tabStack



    fun onBack(){
        println("GO BACK TO HOMESCREEN 1 == ${_fullscreenStack.active}")
        mainNavigation.pop()
        println("GO BACK TO HOMESCREEN 2 == ${_fullscreenStack.active}")
    }
    fun onTabSelected(tab: TabComponent.Tab){
        println("Tab selected == $tab")
        tabNavigation.navigate { stack ->
            stack.filterNot { it.tab == tab } + TabConfig(tab)
        }
    }


    private fun createFullScreenChild(config: MainConfig, componentContext: ComponentContext): FullScreenChild {
        return when(config){
            is MainConfig.None -> FullScreenChild.None
            is MainConfig.PostScreen -> FullScreenChild.PostScreen(postDetailComponent(config,componentContext))
        }
    }

    private fun postDetailComponent(config: MainConfig.PostScreen,componentContext: ComponentContext): PostDetailComponent {
        return PostDetailComponent(
            config.postID,
            componentContext = componentContext,
            onGoBack = { println("Go Back to HomeScreen Component ===") }
        )
    }
    private fun createTab(config: TabConfig, componentContext: ComponentContext): TabComponent {
        return tabFactory.create(
            componentContext = componentContext,
            tab = config.tab,
            onNavigatePost = {
                println("AT GRAND PARENT NOW, EXECUTE EVENT")
                mainNavigation.push(MainConfig.PostScreen(""))
            }
        )
    }


    class Factory(
        private val tabFactory: TabComponent.Factory
    ){
        fun create(
            componentContext: ComponentContext,
            onNavigateTo: (String) -> Unit,
        ) = HomeScreenComponent(
            componentContext = componentContext,
            onNavigateTo = onNavigateTo,
            tabFactory = tabFactory
        )
    }



    sealed interface FullScreenChild {
        object None: FullScreenChild
        class PostScreen(val postDetailsComponent: PostDetailComponent): FullScreenChild
    }
    @Serializable
    sealed interface MainConfig: Parcelable {
        @Serializable
        data object None: MainConfig
        @Serializable
        data class PostScreen(val postID: String): MainConfig
    }

    @Serializable
    private data class TabConfig(val tab: TabComponent.Tab): Parcelable
}

**Reply by arkivanov** [reply]
Thanks! Predictive back gesture on iOS requires some additional setup. Did you follow the [guide](https://arkivanov.github.io/Decompose/extensions/compose/#predictive-back-gesture-on-other-platforms)?

Also probably unrelated, but `val backDispatcher : BackDispatcher = (backHandler as? BackDispatcher) ?: BackDispatcher()` is incorrect. There is no access to `BackDispatcher` in a component, only `BackHandler` is available.

**Reply by OsmanMusse** [reply]
@arkivanov Thanks, will have a look, just wanted to add also how would I be able to access  PredictiveBackGestureOverlay doesn't seem to be in com.arkivanov.decompose.extensions.compose.jetbrains

**Reply by arkivanov** [reply]
It should be there, available for all non-Android targets. See: https://github.com/arkivanov/Decompose/tree/master/extensions-compose-jetbrains/src/nonAndroidMain/kotlin/com/arkivanov/decompose/extensions/compose/jetbrains

**Reply by OsmanMusse** [reply]
I thought the gesture swiping feature came with predictiveBackAnimation function

**Reply by arkivanov** [reply]
Yes. But the overlay is required on non-Android platforms to dispatch touch events to BackDispatcher. The. `predictiveBackAnimation` just listens to the touch events via BackHandler.

**Reply by OsmanMusse** [reply]
Is PredictiveBackGestureOverlay something required in the root component or in every component

**Reply by arkivanov** [reply]
Only for the root. It's a transparent overlay shown on top of the entire content. See the sample: 
https://github.com/arkivanov/Decompose/blob/v3.0/sample/shared/compose/src/iosMain/kotlin/com/arkivanov/sample/shared/root/RootViewController.kt

**Reply by OsmanMusse** [reply]
Thanks for replying, I've made adjustments doesn't seem to allow to drag from the second screen(white screen with button) back to the home screen, only allows when I pop it via clicking the blue button. Here's my updated BottomNavigation

 
 
     @OptIn(ExperimentalCupertinoApi::class, ExperimentalDecomposeApi::class)
     @Composable
    fun BottomNavigationView(
     component: HomeScreenComponent,
     scope: CoroutineScope,
     bottomSheetState: CupertinoSheetState,
     Children(
         stack = component.fullScreenStack,
         modifier = Modifier.fillMaxSize(),
         animation = backAnimation(
             backHandler = component.backHandler,
             onBack = {
                 println("DO SOMETHING")
             }
         ),
     ) {
         when(it.instance) {
              is HomeScreenComponent.FullScreenChild.None -> {
                  CupertinoBottomSheetScaffold(
                      modifier = Modifier.fillMaxSize(),
                      scaffoldState = CupertinoBottomSheetScaffoldState(bottomSheetState),
                      bottomBar = {
                          val grayTitleColor = Color(158, 158, 158)
                          val grayIconColor = Color(190, 190, 190)
                          val primaryColor = Color(72, 134, 255)

                          val tabStack by component.tabStack.subscribeAsState()

                          CompositionLocalProvider(LocalRippleTheme provides NoRippleTheme) {
                              CupertinoNavigationBar(isTranslucent = false) {
                                  TabComponent.Tab.values().forEach { specificTab ->
                                      val (title, icon) = when (specificTab) {
                                          TabComponent.Tab.Home -> "Home" to MR.images.home_icon_lineal
                                          TabComponent.Tab.Account -> "Account" to MR.images.account_icon
                                          TabComponent.Tab.Post -> "Post" to MR.images.camera_icon_linear
                                          TabComponent.Tab.Saved -> "Saved" to MR.images.favourite_icon
                                          TabComponent.Tab.Settings -> "Settings" to MR.images.setting_icon_linear
                                      }

                                      val bottomTabsOffset = when (specificTab) {
                                          TabComponent.Tab.Account, TabComponent.Tab.Post -> 4.dp
                                          TabComponent.Tab.Saved, TabComponent.Tab.Settings -> 3.dp
                                          else -> 4.5.dp
                                      }

                                      NavigationBarItem(
                                          icon = {
                                              Icon(
                                                  painter = painterResource(icon),
                                                  contentDescription = null,
                                                  modifier = Modifier.fillMaxSize(if (specificTab == TabComponent.Tab.Saved) 0.45F else 0.50F)
                                                      .offset(y = bottomTabsOffset),
                                                  tint = Color.Unspecified,
                                              )
                                          },
                                          onClick = {
                                              component.onTabSelected(tab = specificTab)
                                          },
                                          selected = false,
                                          label = {
                                              Text(
                                                  text = "${title}",
                                                  fontSize = 12.5.sp,
                                                  color = if (tabStack.active.instance.tab == specificTab) primaryColor else grayTitleColor
                                              )
                                          }
                                      )

                                  }

                              }
                          }
                      },
                      sheetDragHandle = {},
                      sheetSwipeEnabled = false,
                      sheetContent = {
                          CupertinoBottomSheetContent {
                              CupertinoButton(
                                  modifier = Modifier.padding(top = 50.dp),
                                  onClick = {
                                      scope.launch {
                                          bottomSheetState.hide()
                                      }

                                  }
                              ) {
                                  Text(
                                      text = "Hello world",
                                  )
                              }
                          }

                      },
                  ) {
                      Children(
                          stack = component.tabStack,
                          modifier = Modifier,
                      ) { child ->
                          TabView(child.instance)
                      }
                  }
              }
              is HomeScreenComponent.FullScreenChild.PostScreen ->
                  CupertinoScaffold(modifier = Modifier.fillMaxSize().background(Color.Blue)){
                      CupertinoButton(
                          modifier = Modifier.offset(y = 50.dp).background(Color.Blue),
                          onClick = {
                              println("BACK BUTTON CLICKED ===")
                              component.onBack()
                          }
                      ){
                          Text("Go Back",color = Color.White)
                      }
                  }

              }
         }

     }

**Reply by arkivanov** [reply]
I don't see any issues currently. Please check the [sample on v3.0 branch](https://github.com/arkivanov/Decompose/tree/v3.0/sample), as it contains app-ios-compose with Compose for iOS and back gesture support. If you still have any issues, please provide a reproducer project that I could run. Thanks!

**Reply by OsmanMusse** [reply]
Here's my link to my repo https://github.com/OsmanMusse/Iguhallee_IOS/. Many thanks!

**Reply by arkivanov** [reply]
Thanks for the reproducer. However, I couldn't build it locally.

```
error: Build input file cannot be found: '/Users/arkivanov/dev/workspace/GoogleService-Info.plist'. Did you forget to declare this file as an output of a script phase or custom build rule which produces it? (in target 'iosApp' from project 'iosApp')
```

It would be nice to have a compiling reproducer without any API involved.

Also, I can see that your `MainViewController` has many issues. Specifically, the root component context and the root component should be created outside of Compose (i.e. outside of `ComposeUIViewController {}`). Could you please follow the [docs](https://arkivanov.github.io/Decompose/extensions/compose/#predictive-back-gesture-on-other-platforms) and the [sample](https://github.com/arkivanov/Decompose/blob/v3.0/sample/shared/compose/src/iosMain/kotlin/com/arkivanov/sample/shared/root/RootViewController.kt)?

**Reply by OsmanMusse** [reply]
understood, how about if I need the backAnimation not in the root but for example HomeScreenComponent

**Reply by arkivanov** [reply]
I think in this case you just specify `predictiveBackAnimation` for `Children(...) {}` at the home level (e.g. in `@Composable fun HomeContent()`), not at the root level. Here is an [example](https://github.com/arkivanov/Confetti/blob/55f8abe4560daccbef8494920eb4c151213090aa/androidApp/src/main/java/dev/johnoreilly/confetti/ui/ConferenceRoute.kt#L30), though in that particular example that's only used on Android (but there should be no difference for iOS).

**Reply by OsmanMusse** [reply]
Hi arkivanov, hope your having a great weekend, Here's my latest reproducer without any API https://github.com/OsmanMusse/Iguhallee_IOS.

Problem: doesn't allow me to drag from going to the post screen via the predictive gesture(blue screen) only works if I click the go back button

**Reply by arkivanov** [reply, marked as answer]
Thanks for the reproducer! It looks like you haven't followed the [guide](https://arkivanov.github.io/Decompose/extensions/compose/#predictive-back-gesture-on-other-platforms) completely. In particular, you missed the step #2.

> Create the root `DefaultComponentContext` and pass the previously created `BackDispatcher` as `backHandler` argument.

Here is the patch that you need to apply:

```diff
Index: composeApp/src/iosMain/kotlin/MainViewController.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/composeApp/src/iosMain/kotlin/MainViewController.kt b/composeApp/src/iosMain/kotlin/MainViewController.kt
--- a/composeApp/src/iosMain/kotlin/MainViewController.kt	(revision 5059f4221a7e963d3a2e0cf333249e66aeadb474)
+++ b/composeApp/src/iosMain/kotlin/MainViewController.kt	(date 1714297199412)
@@ -31,7 +31,7 @@
     ){
         val root = remember {
             DefaultRootComponent(
-                componentContext = DefaultComponentContext(lifecycle = ApplicationLifecycle()),
+                componentContext = DefaultComponentContext(lifecycle = ApplicationLifecycle(), backHandler = backDispatcher),
                 homeScreenFactory = getKoinA.get()
             )
         }
```

**Reply by OsmanMusse** [reply]
Thanks your a star @arkivanov. This fixed the issue can't believe is was a silly mistake by me. Also thanks for replying fast

**Reply by arkivanov** [reply]
Glad that helped!
